DscResources/RegistryAccessEntry/RegistryAccessEntry.psm1

```diff
@@ -0,0 +1,477 @@
+'@
+}
+
+
+Function Get-TargetResource
+{
+	[CmdletBinding()]
+	[OutputType([Hashtable])]
+	param
+	(
+		[parameter(Mandatory = $true)]
+		[System.String]
+		$Path,
+
+        [Parameter(Mandatory=$true)]
+		[Microsoft.Management.Infrastructure.CimInstance[]]
+		$AccessControlList,
+
+		[bool]
+		$Force = $false
+	)
+
+    $NameSpace = "root/Microsoft/Windows/DesiredStateConfiguration"
+
+    if(-not (Test-Path -Path $Path))
+    {
+        $errorMessage = $LocalizedData.ErrorPathNotFound -f $Path
+        throw $errorMessage
+    }
+
+    $currentACL = Get-Acl -Path $Path
+
+    $CimAccessControlList = New-Object -TypeName 'System.Collections.ObjectModel.Collection`1[Microsoft.Management.Infrastructure.CimInstance]'
+
+    foreach($Principal in $AccessControlList)
+    {
+        $CimAccessControlEntries = New-Object -TypeName 'System.Collections.ObjectModel.Collection`1[Microsoft.Management.Infrastructure.CimInstance]'
+
+        $PrincipalName = $Principal.Principal
+        $ForcePrincipal = $Principal.ForcePrincipal
+
+        $Identity = Resolve-Identity -Identity $PrincipalName
+        $currentPrincipalAccess = $currentACL.Access.Where({$_.IdentityReference -eq $Identity.Name})
+        foreach($Access in $currentPrincipalAccess)
+        {
+            $AccessControlType = $Access.AccessControlType.ToString()
+            $Rights = $Access.RegistryRights.ToString().Split(',').Trim()
+            $Inheritance = (Get-RegistryRuleInheritenceName -InheritanceFlag $Access.InheritanceFlags.value__ -PropagationFlag $Access.PropagationFlags.value__).ToString()
+
+            $CimAccessControlEntries += New-CimInstance -ClientOnly -Namespace $NameSpace -ClassName AccessControlEntry -Property @{
+                AccessControlType = $AccessControlType
+                Rights = @($Rights)
+                Inheritance = $Inheritance
+                Ensure = ""
+            }
+        }
+
+        $CimAccessControlList += New-CimInstance -ClientOnly -Namespace $NameSpace -ClassName AccessControlList -Property @{
+                        Principal = $PrincipalName
+                        ForcePrincipal = $ForcePrincipal
+                        AccessControlEntry = [Microsoft.Management.Infrastructure.CimInstance[]]@($CimAccessControlEntries)
+                    }
+    }
+
+    $ReturnValue = @{
+        Force = $Force
+        Path = $Path
+        AccessControlList = $CimAccessControlList
+    }
+
+    return $ReturnValue
+}
+
+Function Set-TargetResource
+{
+	[CmdletBinding()]
+	param
+	(
+		[parameter(Mandatory = $true)]
+		[System.String]
+		$Path,
+
+        [Parameter(Mandatory=$true)]
+		[Microsoft.Management.Infrastructure.CimInstance[]]
+		$AccessControlList,
+
+		[bool]
+		$Force = $false
+	)
+
+    if(-not (Test-Path -Path $Path))
+    {
+        $errorMessage = $LocalizedData.ErrorPathNotFound -f $Path
+        throw $errorMessage
+    }
+
+    $currentAcl = Get-Acl -Path $Path
+    if($null -eq $currentAcl)
```

Comment:
What scenario would the ACL be null?

Reply:
I didn’t realize it till after I wrote the code but if the user doesn’t have permissions. So maybe we should throw an error.

Reply:
I found that throwing errors will cause the whole get-dscconfiguration return to fail. Instead we should provide a null value for AccessControlEntry where ACL can't be found, that way the other configurations for that resource will still return values.  I switched around the logic in the NftsAccessEntry resource to do that.

Reply:
Done